pkg/river/token/builder/builder_defaults_test.go

```diff
@@ -0,0 +1,136 @@
+package builder_test
+
+import (
+	"fmt"
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/parser"
+	"github.com/grafana/agent/pkg/river/token/builder"
+	"github.com/grafana/agent/pkg/river/vm"
+	"github.com/stretchr/testify/require"
+	"reflect"
+	"testing"
+)
+
+const defaultNumber = 123
+
+type OuterStructBlockWithEmptyDefault struct {
+	Inner AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OuterStructBlockWithEmptyDefault) SetToDefault() {
+	*o = OuterStructBlockWithEmptyDefault{Inner: AttrWithDefault{}}
+}
+
+type OuterStructBlockWithMatchingDefault struct {
+	Inner AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OuterStructBlockWithMatchingDefault) SetToDefault() {
+	*o = OuterStructBlockWithMatchingDefault{Inner: AttrWithDefault{Number: defaultNumber}}
+}
+
+type AttrWithDefault struct {
+	Number int `river:"number,attr,optional"`
+}
+
+func (i *AttrWithDefault) SetToDefault() {
+	*i = AttrWithDefault{Number: defaultNumber}
+}
+
+type Attr struct {
+	Number int `river:"number,attr,optional"`
+}
+
+type testCase struct {
+	name  string
+	in    interface{}
+	river string
+}
+
+var testCases = []testCase{
+	{
+		name:  "outer struct block / empty defaults / no value set",
+		in:    OuterStructBlockWithEmptyDefault{},
+		river: ``,
+	},
+	{
+		name: "outer struct block / empty defaults / different value set",
+		in: OuterStructBlockWithEmptyDefault{
+			Inner: AttrWithDefault{Number: 321},
+		},
+		river: `
+			inner {
+				number = 321
+			}
+		`,
+	},
+	{
+		name: "outer struct block / empty defaults / default value set",
+		in: OuterStructBlockWithEmptyDefault{
+			Inner: AttrWithDefault{Number: defaultNumber},
+		},
+		// NOTE: this is correct, because inner block's defaults will be applied
+		river: `
+			inner { }
+		`,
+	},
+	{
+		name:  "outer struct block / matching defaults / no value set",
+		in:    OuterStructBlockWithMatchingDefault{},
+		river: `inner { }`,
+	},
```

Comment:
This test appears to fail, because the value encoded to River does not decode into the same original value:

```
 builder_defaults_test.go:120: 
        	Error Trace:	/Users/piotr/workspace/agent/pkg/river/token/builder/builder_defaults_test.go:120
        	Error:      	Not equal: 
        	            	expected: builder_test.OuterStructBlockWithMatchingDefault{Inner:builder_test.AttrWithDefault{Number:0}}
        	            	actual  : builder_test.OuterStructBlockWithMatchingDefault{Inner:builder_test.AttrWithDefault{Number:123}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,3 +2,3 @@
        	            	  Inner: (builder_test.AttrWithDefault) {
        	            	-  Number: (int) 0
        	            	+  Number: (int) 123
        	            	  }
        	Test:       	TestDefaults/outer_struct_block_/_matching_defaults_/_no_value_set
        	Messages:   	Encoding and decoding the same block gave different results
```